x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/planner/ExpressionTranslators.java

```diff
@@ -391,18 +391,32 @@ public static Query doTranslate(In in, TranslatorHandler handler) {
             if (in.value() instanceof FieldAttribute) {
                 // equality should always be against an exact match (which is important for strings)
                 FieldAttribute fa = (FieldAttribute) in.value();
-                List<Expression> list = in.list();
+                DataType dt = fa.dataType();
 
-                // TODO: this needs to be handled inside the optimizer
-                list.removeIf(e -> DataTypes.isNull(e.dataType()));
-                DataType dt = list.get(0).dataType();
+                List<Expression> list = in.list();
                 Set<Object> set = new LinkedHashSet<>(CollectionUtils.mapSize(list.size()));
-
-                for (Expression e : list) {
-                    set.add(handler.convert(valueOf(e), dt));
+                list.forEach(e -> {
+                    // TODO: this needs to be handled inside the optimizer
+                    if (DataTypes.isNull(e.dataType()) == false) {
+                        set.add(handler.convert(valueOf(e), dt));
+                    }
+                });
+
+                if (dt == DATETIME) {
+                    DateFormatter formatter = DateFormatter.forPattern(DATE_FORMAT);
+
+                    q = null;
+                    for (Object o : set) {
+                        assert o instanceof ZonedDateTime;
+                        // see comment in Ranges#doTranslate() as to why formatting as String is required
+                        String zdt = formatter.format((ZonedDateTime) o);
```

Comment:
@matriv, related to your [suggestion](https://github.com/elastic/elasticsearch/pull/60377#issuecomment-681878439), there are similarities with `BinaryComparisons`'s times values code, but not sure if enough for a common extraction to make sense (despite the comment reference). But happy to further iterate.

Reply:
Yeah, seems find as is.